### PR TITLE
fix: Use helm login for Oci helm repos. #5856

### LINKED ISSUE #5856: Unable to update Helm chart dependencies if there is any Helm OCI repository configured
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

- [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
- [x] I've included steps to reproduce the bug.
- [x] I've pasted the output of `argocd version`.

**Describe the bug**

Argo CD fails to update Helm chart dependencies if there is any Helm OCI repository configured in the system.
This OCI repository may not be even used, it just needs to be added to Argo CD to cause the issue. 

It looks like Argo CD is trying to execute `helm repo add ...` command for every registered Helm repository before running `helm dependency update`. If there is Helm OCI repository registered in Argo CD, `helm repo add ...` fails for such OCI repo.

`time="2021-03-24T15:19:21-04:00" level=fatal msg="rpc error: code = InvalidArgument desc = application spec is invalid: InvalidSpecError: Unable to generate manifests in application-chart: rpc error: code = Unknown desc = ``helm repo add --username ****** --password ****** test-helm-oci-repo myregistry.azurecr.io/helm`` failed exit status 1: Error: could not find protocol handler for:`

As far as I understand `helm repo add ...` doesn't work with Helm OCI repos, so probably Helm OCI repos should be excluded from `helm repo add ...`  step.

**To Reproduce**

Please refer to the sample project: https://github.com/lsolovey/argocd-helm-oci

1) Add any Helm OCI repository. It’s not going to be used by our application, it just needs to be added to Argo CD:

`argocd repo add myregistry.azurecr.io/helm --type helm --name test-helm-oci-repo --enable-oci --username <username> --password <password>`

2) Create new Application pointing to the Helm chart which has some dependencies declared in **Chart.yaml**

`argocd app create sample-app --dest-namespace argocd --dest-server https://kubernetes.default.svc --repo https://github.com/lsolovey/argocd-helm-oci.git --path application-chart --sync-policy auto` 

Application creation fails with the error mentioned above.

**Expected behavior**

Application should be created successfully.

**Version**

```shell
argocd: v1.8.7+eb3d1fb
  BuildDate: 2021-03-03T07:16:29Z
  GitCommit: eb3d1fb84b9b77cdffd70b14c4f949f1c64a9416
  GitTreeState: clean
  GoVersion: go1.14.12
  Compiler: gc
  Platform: windows/amd64
argocd-server: v1.8.7+eb3d1fb
  BuildDate: 2021-03-03T07:13:34Z
  GitCommit: eb3d1fb84b9b77cdffd70b14c4f949f1c64a9416
  GitTreeState: clean
  GoVersion: go1.14.12
  Compiler: gc
  Platform: linux/amd64
  Ksonnet Version: v0.13.1
  Kustomize Version: v3.8.1 2020-07-16T00:58:46Z
  Helm Version: v3.4.1+gc4e7485
  Kubectl Version: v1.17.8
  Jsonnet Version: v0.17.0
```

### PR DESCRIPTION
**Pull request by AustinSmart**
Signed-off-by: Austin Smart <austin@austinsmart.com>

Fixes #5856 

If anyone knows of a publicly available (and reliable) Oci helm repo, I can add a test [here](https://github.com/argoproj/argo-cd/blob/1c318f39598bfe2939ca11d70a9b66fd28ec82f7/util/helm/helm_test.go#L105)

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by alexmt** [comment]
You can run OCI registry locally: https://helm.sh/docs/topics/registries/#running-a-registry . Another option is to use Harbor demo server: https://goharbor.io/docs/2.2.0/install-config/demo-server/

**Comment by AustinSmart** [comment]
> You can run OCI registry locally: https://helm.sh/docs/topics/registries/#running-a-registry . Another option is to use Harbor demo server: https://goharbor.io/docs/2.2.0/install-config/demo-server/

Would this fit more in the e2e tests scope than a unit test?

**Comment by alexmt** [comment]
It fits into e2e tests but unfortunately, we don't have the infrastructure to run the local OCI registry yet. I would make changes and test manually, for now, to finally support OCI and add e2e tests as the next step.

**Comment by AustinSmart** [comment]
I have a private OCI registry that I have validated this PR against (the same one used to confirm the bug [here](https://github.com/argoproj/argo-cd/issues/5856#issuecomment-806243137))

**Comment by alexmt** [comment]
We've tried to use private Github repos but it makes CI very fragile. Instead, we now running local git http/ssl and helm servers. It should not be extremely difficult to add an OCI registry too. I would prefer this way.

File: util/helm/helm.go [hunk 1]
```diff
@@ -67,10 +68,26 @@ func (h *helm) Template(templateOpts *TemplateOpts) (string, error) {
...
+			defer func() {
+				h.cmd.IsHelmOci = true
+				_, _ = h.cmd.Logout(repo.Repo, repo.Creds)
+				h.cmd.IsHelmOci = false
+			}()
+
+			if err != nil {
+				return err
+			}
+		} else {
```
Comment (alexmt) [inline util/helm/helm.go:85]: Why is `cmd.RepoAdd` skipped for OCI repositories? I think it still has to be done repo OCI repo can be accessed  via alias, is not it?

**Review by alexmt — CHANGES_REQUESTED** [review]

File: util/helm/helm.go (hunk 1 above)
Comment (AustinSmart) [reply util/helm/helm.go:85]: This is my understanding of helm's current support for OCI registries:

OCI registries do not have an alias. The URL of the registry/repo is what is used to reference charts when pushing/pulling from an OCI registry. 

The only `helm` CLI sub commands that support communicating with an OCI registry are:
`helm registry (login|logout)`
`helm chart (push|pull)`
`helm dependency (build|list|update)`

Running `helm registry login` makes the repo available locally (authenticated) in the same way `helm repo add --username --password` does for non-OCI repos.

Any charts that have a dependency on an OCI repo need to reference the subchart correctly by [prefixing the chart ref ](https://github.com/helm/helm/blob/213a7df2dcd07e8d49fc6acf8a53a540c661ed53/pkg/downloader/manager.go#L398)with `oci://`. Which in turn triggers helm to [use the registry client](https://github.com/helm/helm/blob/213a7df2dcd07e8d49fc6acf8a53a540c661ed53/pkg/action/pull.go#L92) when pulling charts

Comment (alexmt) [reply util/helm/helm.go:85]: thank you for explanation!

**Review by alexmt — APPROVED** [review]
LGTM

**Comment by suhasgaddam-trueaccord** [comment]
@alexmt would it be possible the cherry pick #5887 and #5888 in to `release-1.8` and release a new version of v1.8?

(I don't know what the normal process is.)

**Pull request merged**